redis/src/cluster_routing.rs

```diff
@@ -336,32 +336,144 @@ impl ResponsePolicy {
     }
 }
 
+enum RouteBy {
+    AllNodes,
+    AllPrimaries,
+    FirstKey,
+    MultiShardNoValues,
+    MultiShardWithValues,
+    Random,
+    SecondArg,
+    SecondArgSlot,
+    StreamsIndex,
+    ThirdArgAfterKeyCount,
+    Undefined,
+}
+
+fn base_routing(cmd: &[u8]) -> RouteBy {
+    match cmd {
+        b"ACL SETUSER" | b"ACL DELUSER" | b"ACL SAVE" | b"CLIENT SETNAME" | b"CLIENT SETINFO"
+        | b"SLOWLOG GET" | b"SLOWLOG LEN" | b"SLOWLOG RESET" | b"CONFIG SET"
+        | b"CONFIG RESETSTAT" | b"CONFIG REWRITE" | b"SCRIPT FLUSH" | b"SCRIPT LOAD"
+        | b"LATENCY RESET" | b"LATENCY GRAPH" | b"LATENCY HISTOGRAM" | b"LATENCY HISTORY"
+        | b"LATENCY DOCTOR" | b"LATENCY LATEST" => RouteBy::AllNodes,
+
+        b"DBSIZE"
+        | b"FLUSHALL"
+        | b"FLUSHDB"
+        | b"FUNCTION DELETE"
+        | b"FUNCTION FLUSH"
+        | b"FUNCTION KILL"
+        | b"FUNCTION LOAD"
+        | b"FUNCTION RESTORE"
+        | b"FUNCTION STATS"
+        | b"INFO"
+        | b"KEYS"
+        | b"MEMORY DOCTOR"
+        | b"MEMORY MALLOC-STATS"
+        | b"MEMORY PURGE"
+        | b"MEMORY STATS"
+        | b"PING"
+        | b"SCRIPT EXISTS"
+        | b"SCRIPT KILL"
+        | b"WAIT"
+        | b"RANDOMKEY" => RouteBy::AllPrimaries,
+
+        b"MGET" | b"DEL" | b"EXISTS" | b"UNLINK" | b"TOUCH" => RouteBy::MultiShardNoValues,
+        b"MSET" => RouteBy::MultiShardWithValues,
+
+        // TODO - special handling - b"SCAN"
+        b"SCAN" | b"SHUTDOWN" | b"SLAVEOF" | b"REPLICAOF" | b"MOVE" | b"BITOP" => {
+            RouteBy::Undefined
+        }
+
+        b"EVALSHA" | b"EVAL" => RouteBy::ThirdArgAfterKeyCount,
+
+        b"XGROUP CREATE"
+        | b"XGROUP CREATECONSUMER"
+        | b"XGROUP DELCONSUMER"
+        | b"XGROUP DESTROY"
+        | b"XGROUP SETID"
+        | b"XINFO CONSUMERS"
+        | b"XINFO GROUPS"
+        | b"XINFO STREAM" => RouteBy::SecondArg,
+
+        b"XREAD" | b"XREADGROUP" => RouteBy::StreamsIndex,
+
+        // keyless commands with more arguments, whose arguments might be wrongly taken to be keys.
+        // TODO - double check these, in order to find better ways to route some of them.
+        b"ACL DRYRUN"
+        | b"ACL GENPASS"
+        | b"ACL GETUSER"
+        | b"ACL HELP"
+        | b"ACL LIST"
+        | b"ACL LOG"
+        | b"ACL USERS"
+        | b"ACL WHOAMI"
+        | b"AUTH"
+        | b"BGSAVE"
+        | b"CLIENT GETNAME"
```

Comment:
maybe add a todo to change the client routing based on the response here:
https://github.com/redis/redis/issues/13064